src/analyses/variable-sensitivity/abstract_object.cpp

```diff
@@ -188,7 +189,19 @@ abstract_object_pointert abstract_objectt::expression_transform(
     }
   }
 
-  exprt simplified=simplify_expr(constant_replaced_expr, ns);
+  exprt simplified = simplify_expr(constant_replaced_expr, ns);
+
+  for(const exprt &op : simplified.operands())
+  {
+    abstract_object_pointert lhs_abstract_object = environment.eval(op, ns);
+    const exprt &lhs_value = lhs_abstract_object->to_constant();
+    if(lhs_value.is_nil())
+    {
+      return environment.abstract_object_factory(
+        simplified.type(), ns, true, false);
+    }
+  }
```

Comment:
I don't have enough insight, so please forgive if that's a dumb question: you evaluate all the operands (after simplifying the expression); if one of them doesn't evaluate to a constant then you bail out; if, however, all are successful then all the evaluation results seem to be forgotten -- why is this the right thing to do?

Reply:
If all of them evaluate to a constant (constant abstract object actually, which seems to be different from a constant but returns a constant_exprt anyway) then we get to line 205, which is
`return environment.abstract_object_factory(simplified.type(), simplified, ns);` i.e. the evaluation isn't thrown away. This doesn't show up in the diff because this line hasn't changed.

Reply:
@tautschnig some context : this is the default transformer for an abstract domain.  It relies on the fact that the abstract objects interface has a to_constant method.  The logic previously was:

If the abstract domains for each of the children represent constants / singletons, then constant fold (via simplify_expr) and then build the appropriate abstract object to represent that constant.

It is now:

Using the abstract domains for children which represent constants / singletons, if the expression can constant fold, then build the appropriate abstract object to represent that constant.

The net effect is that the abstract interpreter should be able to handle:

```
int x = nondet_int();
assert(x * 0 == 0);
```

Reply:
Thanks both for the elaboration, but what I still don't get is that you evaluate all the operands of the simplified expression, but never use that evaluation result other than checking whether it is constant or not.

Let me try an example:
````c
int x = 0;
int y = 1;
assert(x * y == 0);
````

In this case, `x * y` would be left unchanged by the simplifier, but the abstract domain may report that both are constant (via constant propagation). My (likely flawed) understanding of the code is that it would still return `x * y` when doing `return environment.abstract_object_factory(simplified.type(), simplified, ns);`